### ISSUE TITLE
Using SpatialDropout2D() with dtype'float16' input will trigger unexpected error on CNTK and Theano, but Tensorflow can run well

### ISSUE THREAD

**Issue by Jup11ter** [keras-team/keras#13823]
**System information**

- Have I written custom code (as opposed to using example directory):
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows 10
- Theano backend (yes / no): yes
- Theano version: 1.0.4
- Keras version: 2.3.1
- Python version: 3.6.9
- CUDA/cuDNN version: -
- GPU model and memory: -

**Describe the current behavior**
When the input dtype is float16 and Theano is used as the backend, SpatialDropout2D() will report an error `TypeError('The two branches should have identical types, but they are TensorType(float32, 4D) and TensorType(float16, 4D) respectively. This error could be raised if for example you provided a one element list on the then branch but a tensor on the else branch.',)` in `theano\ifelse.py line 375`.
**But CNTK and Tensorflow can build the model normally.** When this model is used to process input, CNTK will report `<class 'RuntimeError'>, RuntimeError("Node 'Dropout5' (Dropout operation): TensorView: Attempted to create a TensorView<ElemType> on a storage object of a different ElemType.` .**Only tensorflow can run normally**. The above errors are all about dtype mismatch. When I change the input dtype to float32 or float64, all three backends can process the input normally, and have quite similar output (the sum of absolute values of errors is less than 1e-5).
![pic2](https://user-images.githubusercontent.com/61384431/75115643-643f2180-569b-11ea-833f-3b24c421755a.jpg)

Kears's document does not seem to mention the hidden dangers of float16 type input when using some backends. And whether it is cntk or theano, they have a lot of APIs that can still accept float16 as input. **Such undocumented and uncertain errors may cause confusion for users**.

**Code to reproduce the issue**

```python
import os
import numpy as np
import keras
from keras.engine import Model, Input
import keras.backend as K
import importlib


def set_keras_backend(backend='tensorflow'):
    if K.backend() != backend:
        os.environ['KERAS_BACKEND'] = backend
        importlib.reload(K.load_backend)
        importlib.reload(K)
        assert K.backend() == backend


kwargs={'rate': 0.6290002800171914, 'data_format': 'channels_first'}
input_data = (10 * np.random.random((1,32,32,16)))
input = input_data.astype(np.float16)#you can change here to see how the dtyep effect the layer fucntion
set_keras_backend('theano')
K.set_floatx(str(input.dtype))
layer = keras.layers.core.SpatialDropout2D(**kwargs)
inputx = Input(batch_shape=input.shape)
layerx = layer(inputx)
bk_model = Model(inputx, layerx)
model_path = os.path.join(tmppath, 'model.h5')
bk_model.save(model_path, bk_model)

model = load_model(tmppath)
output = model.predict(input)
print('finish')
```

**Comment by Saduf2019** [comment]
@Jup11ter 
Moving this issue to closed status as there has been no activity, in case you still face the error please create a new issue.
